### ISSUE TITLE
put method? 

### ISSUE THREAD

**Issue by tanninettt** [nus-cs2030/2324-s2#266]
would it be safe to say that the put method works in the same way as a .add method that we usually use for ImLists?

**Comment by HaoZeHuin** [comment]
Yes, your right. The put method will add a key-value pair to the pre-existing ImmutableMap and return a new ImmutableMap.

**Comment by aeonsparkz** [comment]
I think there is a difference, put removes the previous value of the key-value and inputs its own value into the key-value, for ImList will just add it in but put will replace if theres repeated keys.

**Comment by brianpang02** [comment]
the put method for the immutable map is slightly different as it will update existing pairs with the same key but different values. but both does return a new object, hence immutable

**Comment by yishao17** [comment]
It works similarly, but if there is an existing key already, it will update the value. but if not, it will add in a new key value pair.  ImList simply adds things in without checking.

**Comment by juliannxe** [comment]
More on the differences in general, ImList doesn't need an iterator like ImmutableMap does (especially since this lab bans keySet())

**Comment by lohkeyyy** [comment]
I think its good to understand how the equals() and hashCode() methods of object works to better understand the put method for ImmutableMap. You are right to say that the put method updates existing pairs with the same key, however we can decide what "same" is. In other words, we can decide when 2 instances of an object are equal by overriding the equals and hashmethod

**Comment by ZHD1987E** [comment]
@tanninettt mind if you close this issue since lab 4 is over for quite some time?
